gateway-server/src/main/java/org/apache/knox/gateway/GatewayServer.java

```diff
@@ -962,10 +985,32 @@ private void processApplicationPathAliases(File warDir, Topology topology) {
     });
   }
 
+  private void addInactiveTopology(final String topologyName) {
+    synchronized (inactiveTopologies) {
+      inactiveTopologies.add(topologyName);
+    }
+  }
+
+  private void removeInactiveTopology(final String topologyName) {
+    synchronized (inactiveTopologies) {
+      inactiveTopologies.remove(topologyName);
+    }
+  }
+
+  private boolean isInactiveTopology(final String topologyName) {
+    boolean result = false;
+    synchronized (inactiveTopologies) {
+      result = inactiveTopologies.contains(topologyName);
+    }
+    return result;
+  }
+
   private synchronized void internalDeactivateTopology( Topology topology ) {
```

Comment:
Does this get called for the first deployment on the initial start as well?
When does it get marked as active again? Does it incorporate discovery at all?
Should this be done outside of the TopologyService? Seems like state that should be managed by that service maybe.

Reply:
Automated testing will be tricky given the small window of time when this is an issue. I'm open to suggestions though.

When you ask "Does this get called for the first deployment...", what is "this"? On first deployment, there had not been any opportunity for it to have been marked as inactive, so I think the answer is "no".

The topology gets marked active upon activation (see internalActivateTopology method). This is not connected to discovery because we're dealing directly with topologies here. Discovery will have happened prior to this, and the undeployment won't happen until the new version of the topology is written out.

I thought about that, but

1. The source of the 404 responses is Jetty, so overriding them seems appropriate to do with Jetty facilities.
2. It seems best to do it as close to the activation/deactivation logic as possible.

Reply:
"On first deployment, there had not been any opportunity for it to have been marked as inactive, so I think the answer is "no"."

If we don't automatically call deactivate prior to deployment and activate then 404's will still be available while topologies are being deployed. No?

I think the state and isTopologyActive to see if it is active or not can be in the TopologyService and used in the GatewayService error handler. Spreading that state out is what I am questioning.

Reply:
I'm not convinced that the state should be part of the TopologyService. The TopologyService has more to do with handling topology artifacts than actual deployment/activation. It seems to me that the responsibility for deployment/activation is handled by GatewayServer. I think managing this active state in the TopologyService would broaden the window of time during which clients would unnecessarily receive a 503 response where they would have previously not received a 404 response. The 404 responses would not occur until the app was actually deactivated.

Reply:
Concerning the point about initial deployments, we could/should immediately mark topologies as inactive on deployment initiation, and then mark them as active upon completion. I believe this would further minimize the potential for clients to receive a 404 response for a "deployed" topology.

Reply:
I've added the initial deployment behavior you've suggested.